etcd: don't confuse prefixes during migration

awly opened the pull request:
The prefix fetching logic has a bug: it treats everything starting with
`/teleport` as the legacy prefix data, even if it's `/teleport-foo/bar`.
This is an issue if user specifies `/teleport-foo` as their custom
prefix. Each restart will copy the data from `/teleport-foo/...` to
`/teleport-foo-foo/...`.

Set the legacy prefix const to `/teleport/` instead. This avoids
excessive copying during startup.

Fixes https://github.com/gravitational/teleport/issues/4312

Commit:
etcd: don't confuse prefixes during migration

The prefix fetching logic has a bug: it treats everything starting with
`/teleport` as the legacy prefix data, even if it's `/teleport-foo/bar`.
This is an issue if user specifies `/teleport-foo` as their custom
prefix. Each restart will copy the data from `/teleport-foo/...` to
`/teleport-foo-foo/...`.

Set the legacy prefix const to `/teleport/` instead. This avoids
excessive copying during startup.

Prefixes can still be confused later on, with `Watch` and `GetRange`,
but this is harder to migrate with backwards-compatibility.

russjones commented:
@awly Can we make a backup of everything under `/teleport/` before we kick off the migration?

awly commented:
@russjones we can probably backup the data we delete, under the custom prefix (not `/teleport/`).
I'll add that.

Commit:
etcd: backup data before overwrite during migration

In case the migration kicks in by mistake, or ends up deleting more data
than expected, a backup will be very handy.

awly commented:
@russjones PTAL

Commit:
etcd: more tweaks with slashes for migration

When custom prefix is `/foo/`, the migration would move `/teleport/a` to
`/foo/a`, but the backend later tries to read `/foo//a`.

Also added tests to cover these edge cases.

fspmarshall reviewed (approved):
I don't like the idea of having the logic be sensitive to the differences in proceeding/trailing slashes.  Seems like an unnecessary footgun.  Ideally `prefix`, `/prefix`, and `/prefix/` would all mean the same thing from the user's perspective.

Otherwise, LGTM.

Commit:
etcd: fix data race in migration tests

awly commented:
@fspmarshall yeah, I agree.
We made `/prefix` and `prefix` the same, but not with trailing slashes. Now some users have data under `/prefix//foo` so we'll need to be careful when fixing this. That might need another migration :(

benarent commented:
When we update the docs, I'll come back and update a setup guide for etcd, this does seem to follow the slightly confusing prefix logic of etcd as Forrest has pointed out https://etcd.io/docs/v3.3.12/op-guide/authentication/

btw, we should update https://github.com/gravitational/teleport/issues/2883 to make it clear that it needs the trailing slash incase anyone else finds that ticket. ( or is this only an issue during a migration )

fspmarshall commented:
> We made `/prefix` and `prefix` the same, but not with trailing slashes. Now some users have data under `/prefix//foo` so we'll need to be careful when fixing this. That might need another migration :(

Huh.  That *is* tricky. Its too bad we don't have a more well-defined format for the keys used by the `Backend` interface.  It would be nice to be able to unambiguously separate path sections.

I suppose that while `/prefix//foo` looks a bit silly, at least it is unambiguous :shrug:

awly commented:
@benarent actually, all our backed keys start with a `/` so the trailing slash in prefix is not needed.
Updated the description.

Still need approval from @webvictim or @russjones for `/examples/`

russjones commented on lib/backend/etcdbk/etcd.go:
```diff
@@ -687,19 +692,40 @@ func (b *EtcdBackend) syncLegacyPrefix(ctx context.Context) error {
+		key := backupPrefix + strings.TrimPrefix(string(kv.Key), b.cfg.Key)
+		b.Debugf("Copying %q -> %q", kv.Key, key)
+		if _, err := b.client.Put(ctx, key, string(kv.Value)); err != nil {
+			errs = append(errs, trace.WrapWithMessage(err, "failed copying %q to %q: %v", kv.Key, key, err))
```
Shouldn't we stop migration if this occurs?

awly replied:
should we crash the teleport process then, or let it proceed with stale/missing data in custom prefix?

awly replied:
crashing makes sense, it means we can't write to etcd.
updated.
